Handle Etag generation in one place

iilyak opened the pull request:
COUCHDB-2650

iilyak-ibm committed:
Handle Etag generation in one place

COUCHDB-2650

kxepal commented:
Good idea! May be also extract attachment etag generation for consistency?

rnewson commented on src/chttpd.erl:
```diff
@@ -535,7 +535,15 @@ json_body_obj(Httpd) ->
+    Rev = couch_doc:rev_to_str({Start, DiskRev}),
+    <<"\"", Rev/binary, "\"">>.
+
+make_etag() ->
```
a random uuid is essentially useless. The only reason cloudant does that (for views) is to preserve the shape of the code relative to couch. Better to omit the etag header entirely if it isn't going to be stable.

kxepal replied:
Why it was used in CouchDB then? Not because to overcome some cache issues?

iilyak replied:
@rnewson: I don't think changing how `views`  handle etag generation is in the scope of the bug. I could work on it in the context of another bug if you think it needs to be done before 2.0 release. I would approach this by putting `#mrst.sig` into meta. Then when we get [here](https://github.com/apache/couchdb-couch-mrview/blob/master/src/couch_mrview_http.erl#L331), we can use `#mrst.sig` to calculate etag instead of relying on `#vacc.etag` (which is uuid at this point).

iilyak commented:
Should I call `chttpd`'s version of etag related functions from `couch` application in cases [like](https://github.com/apache/couchdb-couch/blob/952a41c7a10fc4dcb0324d09c5be7e5d1d548803/src/couch_httpd_db.erl#L121). Alternatively I can update `couch_chttpd` to dispatch the call to `chttpd`. Which of the above you would prefer?

kxepal commented:
I think not if only you're going to slowly replace couch_httpd by chttpd. Otherwise, it will make more complicated references between subprojects (they are already quite complicated).

iilyak-ibm committed:
Extract `chttpd:att_etag/2`

COUCHDB-2650

iilyak-ibm committed:
(fixup): Wrap base64 encoded Md5 into double quotes

iilyak-ibm committed:
Add headers manipulation functions

COUCHDB-2650

iilyak-ibm committed:
Consider keys in Etag calculation in all_docs_view

COUCHDB-2650

iilyak-ibm committed:
Consider query args in Etag calc (handle_doc_show)

COUCHDB-2650

iilyak-ibm committed:
Use couch_mrview_show:show_etag/4

Remove code duplication by using similar function from couch_mrview_show.
show_etag from couch_mrview_show considers query parameteres so we don't
need to pass `lists:usort(chttpd:qs(Req))` in Extra.

COUCHDB-2650

iilyak-ibm committed:
Calculate Etag for handle_view_list

We used to use uuid here. We do proper calculation of Etag now.

COUCHDB-2650

iilyak-ibm committed:
Consider keys in Etag calc in design_doc_view

COUCHDB-2650

iilyak-ibm committed:
Don't generate Etag for multiquery view

COUCHDB-2650

iilyak-ibm committed:
Don't use pre_flight function for Etag calculation

COUCHDB-2650

davisp commented:
This looks mostly good to me. I'm confused by the changes to add header manipulation functions in 9ab5b83 which don't appear to be used anywhere. Is there a corresponding branch in couch_mrview that's not linked here?

davisp commented:
Ah, found it. Reviewing the others as well.

iilyak commented:
stale PR
